### PR TITLE
ForEach functions

### PR DESCRIPTION
**Pull request by kalaninja**
ForEach pack of functions. These might be really helpful.

### CODE REVIEW

File: result.go [hunk 1]
```diff
@@ -225,6 +225,77 @@ func (q Query) FirstWithT(predicateFn interface{}) interface{} {
 	return q.FirstWith(predicateFunc)
 }
 
+// ForEach performs the specified action on each element of a collection.
+func (q Query) ForEach(action func(interface{})) {
```
Comment (cleitonmarx) [inline result.go]: @kalaninja Awesome! What if we change the action signature to return an error? 
```go
func (q Query) ForEach(action func(interface{}) error) err error {
  next := q.Iterate()
   
  for item, ok := next(); ok; item, ok = next() {
    if err = action(item); err != nil{
      break
    }
  }
  return
}
```
if the action returns an error, ForEach breaks the loop and return the error.

**Comment by ahmetalpbalkan** [comment]
I think we did the right thing by getting rid of errors in function signatures. So I am not a big fan of adding them back. It looks like I can achieve what `ForEach` does for me with `Select`, too, right? Currently, I'm -1 for this proposal:

As an alternative solution, I can have a `SelectT` which is `func(i int) error` my result goes into a `chan error`, then I can process all the errors coming up at the end, right? I won’t be able to terminate on the first error, but it will work pretty much the same?

To take a step back, it looks like `ForEach` is designed to be used with functions with side effects (i.e. querying network/database, printing output etc). Those usually return error types and usually they take long time. In .NET they `.ForEach` very frequently goes together with `.AsParallel`.

Since we don't have `.AsParallel`, I suggest we "show" people how they can use channels more, instead of having a `ForEach`. This way, they can even control the parallelization level on the loop. How messy does the code look like if we try to do `ForEach` with/without errors using SelectT+chan?

File: result.go (hunk 1 above)
Comment (kalaninja) [inline result.go]: yep, that really is a good idea

**Comment by kalaninja** [comment]
>  It looks like I can achieve what ForEach does for me with Select, too, right?

Not really. `Select` returns a query, so you need to iterate it afterwards. So you can substitute `ForEach `with something like
```go
ignoredValue := Select(func(item interface{}) (somethingThatIDontReallyNeed interface{}) {
  action(item)
  somethingThatIDontReallyNeed = true
} ).Results()
```
But using something like 
```go
ForEach(func(item interface{}) {
  action(item)
} )
```
seems better to me.


Though it is not the best case, our "concurrent" example from the site:
```go
urls := make(chan interface{})
results := make(chan interface{})

go func(domains []string) {
	From(domains).Select(func(d interface{}) interface{} {
		return fmt.Sprintf("http://%s/favicon.ico", d)
	}).ToChannel(urls)
}(domains)

go func() {
	defer close(results)
	var wg sync.WaitGroup
	for url := range urls {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			resp, _ := http.Get(url)
			results <- resp.StatusCode
		}(url.(string))
	}
	wg.Wait()
}()

FromChannel(results).CountWith(func(r interface{}) bool {
	return r.(int) == 200
})
```

... can be rewritten with no extra channels and less code with `ForEach`
```go
results := make(chan interface{})

var wg sync.WaitGroup
go func(domains []string) {
	defer close(results)
	From(domains).Select(func(d interface{}) interface{} {
		return fmt.Sprintf("http://%s/favicon.ico", d)
	}).ForEach(func(url interface{}) {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			resp, _ := http.Get(url)
			results <- resp.StatusCode
		}(url.(string))
	})
	wg.Wait()
}(domains)

FromChannel(results).CountWith(func(r interface{}) bool {
	return r.(int) == 200
})
```

**Comment by kalaninja** [comment]
> I think we did the right thing by getting rid of errors in function signatures. So I am not a big fan of adding them back.

I got the point, and agree with you. And if one needs to break the `ForEach` at first error, it can be done with panic.

**Comment by ahmetalpbalkan** [comment]
Thanks! I guess we can close it for now. BTW next time before implementing something feel free to just bring up a proposal, it looks like we frequently kill PRs with code in the design stage. :)

**Comment by kalaninja** [comment]
Sure. 
No luck this time.

**Comment by ahmetalpbalkan** [comment]
Apologies, I was on mobile, so I didn't see your comment with the code examples... It actually makes sense to have:

    ForEach(func(){...
         go job()
    })

We can reopen this back. It think we can just accept `func (v interface{})` without any `error` types. This way, users can decide:

1. how they want to handle early termination / cancellation
2. how they want to handle errors (e.g. one or multiple)
3. they can configure parallelism by forking into background and using WaitGroup etc.

**Comment by kalaninja** [comment]
I am reopening this pr
And I've added some examples to keep test coverage at 100%

File: result_test.go [hunk 2]
```diff
@@ -197,6 +197,60 @@ func TestFirstWithT_PanicWhenPredicateFnIsInvalid(t *testing.T) {
...
+		want  interface{}
+	}{
+		{[5]int{1, 2, 2, 35, 111}, []int{1, 2, 2, 35, 111}},
+		{[]int{}, []int{}},
+	}
+
+	for _, test := range tests {
+		output := []int{}
+		From(test.input).ForEach(func(item interface{}) {
+			output = append(output, item.(int))
```
Comment (ahmetalpbalkan) [inline result_test.go]: maybe make this test `i*2` so that we make sure the predicate is involved :)

File: example_test.go [hunk 3]
```diff
@@ -1735,6 +1735,38 @@ func ExampleQuery_FirstWithT() {
 
 }
 
+// The following code example demonstrates how to use ForEachT
+// to output all elements of an array.
+func ExampleQuery_ForEachT() {
```
Comment (ahmetalpbalkan) [inline example_test.go]: Perhaps add examples for ForEach/ForEachIndexed too?

File: result_test.go (hunk 2 above)
Comment (kalaninja) [inline result_test.go]: yes, sure

File: example_test.go (hunk 3 above)
Comment (kalaninja) [inline example_test.go]: ok

**Comment by kalaninja** [comment]
@ahmetalpbalkan Thanks for the comments. Now it is done.

**Pull request merged**